File: Trees/balanced_binary_tree.py
# https://leetcode.com/problems/balanced-binary-tree

from typing import Optional
from TreeNode import TreeNode


# Time Complexity = O(N)
class Solution:
    def check_balance(self, root):
        if root is None:
            return 0, True

        left = self.check_balance(root.left)
        right = self.check_balance(root.right)
        height = max(left[0], right[0]) + 1

        if abs(left[0] - right[0]) <= 1 and left[1] and right[1]:
            return height, True

        return height, False

    def is_balanced(self, root: Optional[TreeNode]) -> bool:
        res = self.check_balance(root)
        return res[1]

File: Trees/balanced_binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    private Integer maxDepth(TreeNode root){
        if(root==null){
            return 0;
        }

        Integer leftDepth = maxDepth(root.left);
        Integer rightDepth = maxDepth(root.right);

        if (leftDepth == null || rightDepth == null){
            return null;
        }

        if (Math.abs(leftDepth - rightDepth) > 1){
            return null;
        }

        return Math.max(leftDepth, rightDepth) + 1;
    }
    public boolean isBalanced(TreeNode root) {
        return maxDepth(root) != null;
    }
}